# Task 2.3 - Result, Comparison and Reproducibility Checklist

In [1]:
from pathlib import Path
import sys
ROOT = Path.cwd()
if str(ROOT) not in sys.path:
    sys.path.insert(0, str(ROOT))

from submission_support.drm_support import metrics_table, run_main_experiment

result = run_main_experiment()
rows = metrics_table(result["metrics"])
rows

[{'model': 'Simplified DRM',
  'train_accuracy': 0.9103,
  'test_accuracy': 0.9365,
  'train_error': 0.0897,
  'test_error': 0.0635,
  'mean_pairwise_alignment': 0.5616},
 {'model': 'No-diversity ensemble',
  'train_accuracy': 0.9145,
  'test_accuracy': 0.9286,
  'train_error': 0.0855,
  'test_error': 0.0714,
  'mean_pairwise_alignment': 1.0},
 {'model': 'RBF SVM baseline',
  'train_accuracy': 0.9402,
  'test_accuracy': 0.9365,
  'train_error': 0.0598,
  'test_error': 0.0635,
  'mean_pairwise_alignment': 'NA'}]

## Outcome Discussion

                My achieved result for the full simplified DRM model is **test error = 0.0635** on the toy moons dataset. The most comparable paper claim is not a single matching experiment but the paper's recurring observation in Table 1 that DRM can beat a standard SVM on several binary classification datasets; for example, on the **German** dataset the paper reports **DRM21 = 0.280 ± 0.017** versus **SVM = 0.287 ± 0.008**.

                The numbers differ for several honest reasons. First, I used a synthetic toy dataset instead of one of the paper's UCI datasets, so the difficulty, dimensionality, and noise pattern are completely different. Second, my implementation is a simplified DRM-style surrogate with a random Fourier feature map and gradient-based training, not the exact alternating QCQP/dual solver from Eq. (4)-(7). Third, the paper tunes `mu`, `nu`, kernel width, and ensemble size with its own protocol, whereas I used a small reproducible configuration designed to stay CPU-friendly and easy to explain in viva. I therefore treat the performance gap as expected rather than as a failure of reproduction.

                ## Metric Table

                ```text
                model                 | train_accuracy | test_accuracy | train_error | test_error | mean_pairwise_alignment
----------------------+----------------+---------------+-------------+------------+------------------------
Simplified DRM        | 0.9103         | 0.9365        | 0.0897      | 0.0635     | 0.5616                 
No-diversity ensemble | 0.9145         | 0.9286        | 0.0855      | 0.0714     | 1.0                    
RBF SVM baseline      | 0.9402         | 0.9365        | 0.0598      | 0.0635     | NA                     
                ```

## Reproducibility Checklist

- Random seeds are set and documented at the top of each notebook, where applicable.
- All dependencies are listed in `partB/requirements.txt` with version numbers.
- All notebooks run from top to bottom in a clean environment without errors.
- Dataset loading requires no undocumented manual steps because the CSV files are saved under `partB/data/`.
- All hyperparameters are clearly named and defined in one place rather than scattered across cells.